File: android/app/src/main/kotlin/com/ridelink/app/music/IntercomMusicCoexistenceCoordinator.kt
package com.ridelink.app.music

import androidx.annotation.MainThread
import com.ridelink.core.audiopolicy.CoexistenceAction
import com.ridelink.core.audiopolicy.CoexistenceFallback
import com.ridelink.core.audiopolicy.CoexistenceInput
import com.ridelink.core.audiopolicy.CoexistenceState
import com.ridelink.core.audiopolicy.IntercomMusicCoexistence
import com.ridelink.core.audiopolicy.IntercomPolicy
import com.ridelink.core.audiopolicy.RouteState
import com.ridelink.core.player.PlayerState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.math.roundToInt

interface CoexistenceEventSink {
    fun onMusicChanged(state: PlayerState)

    fun onPlaybackIntent(playing: Boolean)

    fun onBaseVolumeChanged(volumePermille: Int)
}

/** Narrow player boundary used by the coexistence owner and its deterministic tests. */
interface MusicCoexistencePort {
    val coexistencePlayerState: StateFlow<PlayerState>
    val coexistenceBaseVolumePermille: StateFlow<Int>

    suspend fun beginCoexistenceLifetime(generation: Long)

    suspend fun applyCoexistenceGain(
        generation: Long,
        volumePermille: Int,
    ): Boolean

    suspend fun pauseForVoice(
        generation: Long,
        trackToken: String,
    ): Boolean

    suspend fun resumeAfterVoice(
        generation: Long,
        trackToken: String,
    ): Boolean

    var coexistenceEvents: CoexistenceEventSink?
}

fun interface GainRampSleeper {
    suspend fun sleep(stepDurationMs: Long)
}

data class CoexistenceDiagnostics(
    val generation: Long = 0,
    val targetVolumePermille: Int = CoexistenceState.FULL_GAIN_PERMILLE,
    val appliedVolumePermille: Int = CoexistenceState.FULL_GAIN_PERMILLE,
    val pausedByVoice: Boolean = false,
    val routeState: RouteState = RouteState.STABLE,
    val fallback: CoexistenceFallback = CoexistenceFallback.NONE,
    val rampRevision: Long = 0,
    val rampCancellationCount: Int = 0,
    val pauseCount: Int = 0,
    val resumeCount: Int = 0,
    val staleInputCount: Int = 0,
)

/**
 * The sole driver of Phase 6 music effects. [IntercomMusicCoexistence] makes every decision; this
 * class serialises its effects and owns the cancellable 200 ms gain ramp.
 */
class IntercomMusicCoexistenceCoordinator(
    private val scope: CoroutineScope,
    private val music: MusicCoexistencePort,
    private val sleeper: GainRampSleeper = GainRampSleeper { delay(it) },
) : CoexistenceEventSink {
    private var state = CoexistenceState()
    private var nextGeneration = 0L
    private var effectTail: Job? = null
    private var rampJob: Job? = null
    private var rampRevision = 0L
    private var appliedVolumePermille = CoexistenceState.FULL_GAIN_PERMILLE
    private var rampCancellationCount = 0
    private var pauseCount = 0
    private var resumeCount = 0

    private val _diagnostics = MutableStateFlow(CoexistenceDiagnostics())
    val diagnostics: StateFlow<CoexistenceDiagnostics> = _diagnostics.asStateFlow()

    init {
        music.coexistenceEvents = this
        onBaseVolumeChanged(music.coexistenceBaseVolumePermille.value)
        onMusicChanged(music.coexistencePlayerState.value)
    }

    @MainThread
    fun beginLifetime(policy: IntercomPolicy): Long {
        val generation = ++nextGeneration
        val outcome = IntercomMusicCoexistence.reduce(state, CoexistenceInput.LifetimeStarted(generation, policy))
        state = outcome.state

        // A successor never waits for predecessor work. Cancellation is backed by the player's own
        // generation check, so an uncooperative old continuation is still inert when it resumes.
        effectTail?.cancel()
        rampJob?.cancel()
        effectTail =
            scope.launch {
                music.beginCoexistenceLifetime(generation)
                perform(generation, outcome.actions)
            }
        publishDiagnostics()
        return generation
    }

    @MainThread
    fun endLifetime(generation: Long) {
        submit(CoexistenceInput.LifetimeEnded(generation))
    }

    /** Terminal lifecycle seam: restoration and any cancelled ramp are complete on return. */
    suspend fun awaitLifetimeEnded() {
        awaitEffectsSettled()
    }

    @MainThread
    fun selectPolicy(
        generation: Long,
        policy: IntercomPolicy,
    ) {
        submit(CoexistenceInput.PolicySelected(generation, policy))
    }

    @MainThread
    fun updateVoice(
        generation: Long,
        available: Boolean,
        localTransmitting: Boolean,
        peerTransmitting: Boolean,
        routeState: RouteState,
        interrupted: Boolean,
        transitionTimedOut: Boolean,
    ) {
        submit(CoexistenceInput.VoiceChanged(generation, available, localTransmitting, peerTransmitting))
        submit(CoexistenceInput.RouteChanged(generation, routeState, interrupted, transitionTimedOut))
    }

    @MainThread
    fun updateSyncAvailability(available: Boolean) {
        submit(CoexistenceInput.SyncAvailabilityChanged(state.generation, available))
    }

    /** Observable completion seam for deterministic lifecycle tests; production never needs a sleep. */
    suspend fun awaitEffectsSettled() {
        effectTail?.join()
        rampJob?.join()
    }

    @MainThread
    override fun onMusicChanged(state: PlayerState) {
        submit(
            CoexistenceInput.MusicChanged(
                generation = this.state.generation,
                available = state.error == null,
                trackToken = state.localEntryId?.value,
                playing = state.playing,
                ended = state.ended,
            ),
        )
    }

    @MainThread
    override fun onPlaybackIntent(playing: Boolean) {
        submit(CoexistenceInput.UserPlaybackIntent(state.generation, playing))
    }

    @MainThread
    override fun onBaseVolumeChanged(volumePermille: Int) {
        submit(CoexistenceInput.BaseVolumeChanged(state.generation, volumePermille))
    }

    @MainThread
    private fun submit(input: CoexistenceInput) {
        val outcome = IntercomMusicCoexistence.reduce(state, input)
        state = outcome.state
        if (outcome.actions.isNotEmpty()) enqueueEffects(state.generation, outcome.actions)
        publishDiagnostics()
    }

    private fun enqueueEffects(
        generation: Long,
        actions: List<CoexistenceAction>,
    ) {
        val previous = effectTail
        effectTail =
            scope.launch {
                previous?.join()
                perform(generation, actions)
            }
    }

    private suspend fun perform(
        generation: Long,
        actions: List<CoexistenceAction>,
    ) {
        for (action in actions) {
            when (action) {
                is CoexistenceAction.RampMusicVolume -> startRamp(generation, action)
                is CoexistenceAction.PauseMusicForVoice -> {
                    if (music.pauseForVoice(generation, action.trackToken)) pauseCount += 1
                }
                is CoexistenceAction.ResumeMusicAfterVoice -> {
                    if (music.resumeAfterVoice(generation, action.trackToken)) resumeCount += 1
                }
            }
        }
        publishDiagnostics()
    }

    private fun startRamp(
        generation: Long,
        action: CoexistenceAction.RampMusicVolume,
    ) {
        if (rampJob?.isActive == true) rampCancellationCount += 1
        rampJob?.cancel()
        val revision = ++rampRevision
        val start = appliedVolumePermille
        val steps = RAMP_STEPS
        val stepDurationMs = action.durationMs / steps
        rampJob =
            scope.launch {
                for (step in 1..steps) {
                    sleeper.sleep(stepDurationMs)
                    if (revision != rampRevision || generation != state.generation) return@launch
                    val fraction = step.toDouble() / steps
                    val value = (start + ((action.targetPermille - start) * fraction)).roundToInt()
                    if (!music.applyCoexistenceGain(generation, value)) return@launch
                    appliedVolumePermille = value
                    publishDiagnostics()
                }
            }
        publishDiagnostics()
    }

    private fun publishDiagnostics() {
        _diagnostics.value =
            CoexistenceDiagnostics(
                generation = state.generation,
                targetVolumePermille = state.targetVolumePermille,
                appliedVolumePermille = appliedVolumePermille,
                pausedByVoice = state.pausedByVoice,
                routeState = state.routeState,
                fallback = state.fallback,
                rampRevision = rampRevision,
                rampCancellationCount = rampCancellationCount,
                pauseCount = pauseCount,
                resumeCount = resumeCount,
                staleInputCount = state.staleInputCount,
            )
    }

    private companion object {
        const val RAMP_STEPS = 10
    }
}

File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Coexistence/IntercomMusicCoexistenceCoordinator.swift
import Foundation
import RideLinkCore

@MainActor
public protocol MusicCoexistencePort: AnyObject {
    var coexistencePlayerState: PlayerState { get }
    var coexistenceBaseVolumePermille: Int { get }
    var coexistenceEvents: (any CoexistenceEventSink)? { get set }

    func beginCoexistenceLifetime(_ generation: Int64) async
    func applyCoexistenceGain(generation: Int64, volumePermille: Int) async -> Bool
    func pauseForVoice(generation: Int64, trackToken: String) async -> Bool
    func resumeAfterVoice(generation: Int64, trackToken: String) async -> Bool
}

@MainActor
public protocol CoexistenceEventSink: AnyObject {
    func onMusicChanged(_ state: PlayerState)
    func onPlaybackIntent(playing: Bool)
    func onBaseVolumeChanged(_ volumePermille: Int)
}

public struct CoexistenceDiagnostics: Sendable, Equatable {
    public var generation: Int64 = 0
    public var targetVolumePermille = CoexistenceState.fullGainPermille
    public var appliedVolumePermille = CoexistenceState.fullGainPermille
    public var pausedByVoice = false
    public var routeState: RouteState = .stable
    public var fallback: CoexistenceFallback = .none
    public var rampRevision: Int64 = 0
    public var rampCancellationCount = 0
    public var pauseCount = 0
    public var resumeCount = 0
    public var staleInputCount = 0

    public init() {}
}

/// The sole iOS driver of Phase 6 music effects. The mirrored pure reducer makes every decision;
/// this type serializes effects and owns the cancellable, generation-bound 200 ms ramp.
@MainActor
public final class IntercomMusicCoexistenceCoordinator: CoexistenceEventSink {
    public typealias Sleeper = @Sendable (UInt64) async -> Void

    public private(set) var diagnostics = CoexistenceDiagnostics()
    private var diagnosticsObserver: (@MainActor (CoexistenceDiagnostics) -> Void)?

    private weak var music: (any MusicCoexistencePort)?
    private let sleeper: Sleeper
    private var state = CoexistenceState()
    private var nextGeneration: Int64 = 0
    private var effectTail: Task<Void, Never>?
    private var rampTask: Task<Void, Never>?
    private var rampRevision: Int64 = 0
    private var appliedVolumePermille = CoexistenceState.fullGainPermille
    private var rampCancellationCount = 0
    private var pauseCount = 0
    private var resumeCount = 0

    public init(
        music: any MusicCoexistencePort,
        sleeper: @escaping Sleeper = { nanoseconds in try? await Task.sleep(nanoseconds: nanoseconds) }
    ) {
        self.music = music
        self.sleeper = sleeper
        music.coexistenceEvents = self
        onBaseVolumeChanged(music.coexistenceBaseVolumePermille)
        onMusicChanged(music.coexistencePlayerState)
    }

    @discardableResult
    public func beginLifetime(policy: IntercomPolicy) -> Int64 {
        nextGeneration += 1
        let generation = nextGeneration
        let outcome = IntercomMusicCoexistence.reduce(state: state, input: .lifetimeStarted(generation: generation, policy: policy))
        state = outcome.state
        effectTail?.cancel()
        rampTask?.cancel()
        effectTail = Task { @MainActor [weak self] in
            guard let self, let music = self.music else { return }
            await music.beginCoexistenceLifetime(generation)
            await self.perform(generation: generation, actions: outcome.actions)
        }
        publishDiagnostics()
        return generation
    }

    public func endLifetime(_ generation: Int64) { submit(.lifetimeEnded(generation: generation)) }

    /// Terminal lifecycle seam: restoration and any cancelled ramp are complete on return.
    public func awaitLifetimeEnded() async { await awaitEffectsSettled() }
    public func selectPolicy(_ policy: IntercomPolicy, generation: Int64) { submit(.policySelected(generation: generation, policy: policy)) }

    public func updateVoice(
        generation: Int64,
        available: Bool,
        localSpeechActive: Bool,
        peerSpeechActive: Bool,
        speechActivityAvailable: Bool,
        routeState: RouteState,
        interrupted: Bool,
        transitionTimedOut: Bool
    ) {
        submit(.voiceChanged(
            generation: generation,
            available: available,
            localSpeechActive: localSpeechActive,
            peerSpeechActive: peerSpeechActive,
            speechActivityAvailable: speechActivityAvailable
        ))
        submit(.routeChanged(generation: generation, routeState: routeState, interrupted: interrupted, transitionTimedOut: transitionTimedOut))
    }

    public func updateSyncAvailability(_ available: Bool) {
        submit(.syncAvailabilityChanged(generation: state.generation, available: available))
    }

    /// Observable completion seam for deterministic lifecycle tests; production never needs a sleep.
    public func awaitEffectsSettled() async {
        _ = await effectTail?.value
        _ = await rampTask?.value
    }

    public func setDiagnosticsObserver(_ observer: @escaping @MainActor (CoexistenceDiagnostics) -> Void) {
        diagnosticsObserver = observer
        observer(diagnostics)
    }

    public func onMusicChanged(_ player: PlayerState) {
        submit(.musicChanged(
            generation: state.generation,
            available: player.error == nil,
            trackToken: player.localEntryId?.value,
            playing: player.playing,
            ended: player.ended
        ))
    }

    public func onPlaybackIntent(playing: Bool) { submit(.userPlaybackIntent(generation: state.generation, playing: playing)) }
    public func onBaseVolumeChanged(_ volumePermille: Int) { submit(.baseVolumeChanged(generation: state.generation, volumePermille: volumePermille)) }

    private func submit(_ input: CoexistenceInput) {
        let outcome = IntercomMusicCoexistence.reduce(state: state, input: input)
        state = outcome.state
        if !outcome.actions.isEmpty { enqueue(generation: state.generation, actions: outcome.actions) }
        publishDiagnostics()
    }

    private func enqueue(generation: Int64, actions: [CoexistenceAction]) {
        let previous = effectTail
        effectTail = Task { @MainActor [weak self] in
            _ = await previous?.value
            guard let self, !Task.isCancelled else { return }
            await self.perform(generation: generation, actions: actions)
        }
    }

    private func perform(generation: Int64, actions: [CoexistenceAction]) async {
        guard let music else { return }
        for action in actions {
            switch action {
            case .rampMusicVolume(let targetPermille, let durationMs):
                startRamp(generation: generation, targetPermille: targetPermille, durationMs: durationMs)
            case .pauseMusicForVoice(let token):
                if await music.pauseForVoice(generation: generation, trackToken: token) { pauseCount += 1 }
            case .resumeMusicAfterVoice(let token):
                if await music.resumeAfterVoice(generation: generation, trackToken: token) { resumeCount += 1 }
            }
        }
        publishDiagnostics()
    }

    private func startRamp(generation: Int64, targetPermille: Int, durationMs: Int64) {
        if rampTask != nil { rampCancellationCount += 1 }
        rampTask?.cancel()
        rampRevision += 1
        let revision = rampRevision
        let start = appliedVolumePermille
        let sleepNs = UInt64(durationMs / Int64(Self.rampSteps)) * 1_000_000
        rampTask = Task { @MainActor [weak self] in
            guard let self else { return }
            for step in 1...Self.rampSteps {
                await sleeper(sleepNs)
                guard !Task.isCancelled, revision == rampRevision, generation == state.generation, let music else { return }
                let fraction = Double(step) / Double(Self.rampSteps)
                let value = Int((Double(start) + Double(targetPermille - start) * fraction).rounded())
                guard await music.applyCoexistenceGain(generation: generation, volumePermille: value) else { return }
                appliedVolumePermille = value
                publishDiagnostics()
            }
        }
        publishDiagnostics()
    }

    private func publishDiagnostics() {
        diagnostics.generation = state.generation
        diagnostics.targetVolumePermille = state.targetVolumePermille
        diagnostics.appliedVolumePermille = appliedVolumePermille
        diagnostics.pausedByVoice = state.pausedByVoice
        diagnostics.routeState = state.routeState
        diagnostics.fallback = state.fallback
        diagnostics.rampRevision = rampRevision
        diagnostics.rampCancellationCount = rampCancellationCount
        diagnostics.pauseCount = pauseCount
        diagnostics.resumeCount = resumeCount
        diagnostics.staleInputCount = state.staleInputCount
        diagnosticsObserver?(diagnostics)
    }

    private static let rampSteps = 10
}
